- Original commit for the rebooted AST

mdedetrich opened the pull request:
Work done due to https://github.com/json4s/json4s/issues/272

mdedetrich committed:
- Original commit for the rebooted AST

rossabaker commented on src/main/scala/org/json4s/ast/JValue.scala:
```diff
@@ -11,36 +10,39 @@ case object JNull extends JValue
-  def apply(field: JField, fields: JField*): JObject = new JObject(field +: fields)
+object JBoolean {
+  @inline def apply(x: Boolean): JBoolean = if (x) JTrue else JFalse
+  @inline def unapply(x: JBoolean): Boolean = x.value
```
Is this the desired extraction idiom?

``` scala
scala> JTrue match { case JBoolean(b) => b }
<console>:13: error: too many patterns for object JBoolean offering Boolean: expected 0, found 1
              JTrue match { case JBoolean(b) => b }
                                 ^

scala> JTrue match { case JBoolean() => true; case _ => false }
res8: Boolean = true
```

I think returning `Option[Boolean]` might be more cohesive with the rest of the extractors, to permit the first syntax.  Those wanting to pass or fail could directly match against `JTrue` or `JFalse`.

mdedetrich replied:
Ah yes, I didn't know the reason behind the `Option[Boolean]`, will fix this

mdedetrich commented:
I will do a benchmark to see if `@inline`, will probably use `ScalaMeter` as a test framework. Next up on the list, is that I will redo the organisation of the `SBT` build definition so that it works with `Scala.js`, and then obviously a bit more work will need to be done to add the relevant info for resolvers/publishing.

Also, regarding the `BigDecimal` formatters, does adding a formatted to `Byte[Array]` and from `Byte[Array]` make sense?. I think it would be useful to establish a stable byte streaming format, but its definitely in the nice to have category

mdedetrich committed:
- Made JBoolean extractor use an Option[Boolean] instead of just Boolean

mdedetrich committed:
- Moved the directory structure around, to account for Scala.js
- Added the Scala.js plugin
- Cleaned up the `build.sbt`, will added in the necessary publish information at the right time
- Added @JSExport's to Scala.js for sane types
- Added extra constructors, only for Scala.js
- Made it cross compile for Scala 2.10.x

mdedetrich committed:
- Added in some of the options from the previous build.sbt
- Need to check the justification for settings like "-target:jvm-1.7"
- Removed the sbt script, no need for this nowadays

mdedetrich committed:
- Added in some of the options from the previous build.sbt
- Removed the sbt script, no need for this nowadays
- Added performance testing through ScalaMeter for JVM backend. For now, just testing against a non inlined version of json-ast, will add other implementations of JSON AST's for comparisons
- Fixed issue with package object not being automatically imported

mdedetrich committed:
- Test suite is now using Microbenchmark so its more accurate

mdedetrich committed:
- Removed JUndefined, there is no compelling use case for it in the AST

rossabaker commented on jvm/src/main/scala/org/json4s/JValue.scala:
```diff
@@ -0,0 +1,48 @@
+
+object JBoolean {
+  @inline def apply(x: Boolean): JBoolean = if (x) JTrue else JFalse
+  @inline def unapply(x: JBoolean): Option[Boolean] = Option(x.value)
```
This could avoid a null check, which probably doesn't matter in the grand scheme, with `Some(x.value)`.

mdedetrich replied:
Well `Option(null)` gives `None` where as `Some(null)` gives `Some(null)`. I think the former is more desirable (also `Some` has some issues with the type system, particularly when working with more complex stuff like `shapeless` or `scalaz`)

rossabaker commented on jvm/src/main/scala/org/json4s/JValue.scala:
```diff
@@ -0,0 +1,48 @@
+package org.json4s
```
The previous version had this in `org.json4s.ast`.  That's a little more verbose, but would allow the json4s reboot to live alongside the legacy version for migration.

I don't have a strong opinion, but a strategy is needed here.

mdedetrich replied:
I will move it to `ast`. I think we will end up having to support both versions anyways.

mdedetrich committed:
- Changed Option to Some, the value can never be null and Some is slightly faster since it wont check for null

mdedetrich committed:
- Moved the implementation to the ast package so it doesn't conflict with current AST
- Very minor syntax adjustments

rossabaker commented:
I have mostly called for subtraction here, but I will argue for a few small additions: `empty` on the companion objects of `JObject`, `JArray`.  Those, I'm sure, are uncontroversial values, and will save a lot of `JObject(Map.empty)` and `JArray(Vector.empty)` calls.

Maybe also `JString.empty` and `JNumber.zero`.

mdedetrich committed:
- Changed JValue from a Trait to a abstract class extends Serializable with Product to help with binary compatibility and type inference

mdedetrich committed:
- Added a readme, is still going to need work
- Fixed up the javascript version of JValue to properly export the correct constructor for use in Javascript

mdedetrich committed:
- Fixed a typo in the README

mdedetrich committed:
- Fixed up phrasing in the document

mdedetrich committed:
- Updated the Readme to provide an explanation of why JSON numbers are unlimited in size

mdedetrich committed:
- Added a String constructor for JNumber in Scala.js, so you can actually create larger larger than 64 bit precision JNumber's in Javascript
- Added documentation about this change

mdedetrich committed:
- Cleaned up the README a bit more

mdedetrich committed:
- Added more links, fixed up the formatting a bit more

mdedetrich committed:
- Mention that the AST is fully immutable

mdedetrich committed:
- Fixed up the tests for InlineTest (so they can at least run)
- Added a test for BigDecimal runtime, in response to measuring performance difference to a potential new unsafe version of json-ast

mdedetrich committed:
- Removed @inline

mdedetrich committed:
- Provided a separate JSON AST implementation at org.json4s.basic.ast. This JSON AST implementation has
 a strict focus on performance, rather than correctness
 - Updated the README to reflect these changes
 - Changed the representation of org.json4s.ast.JArray from immutable.Seq to Vector. If you are concerned about
 memory usage/performance, use org.json4s.basic.ast
 - Renamed the BigDecimalConverter to JNumberConverter for org.json4s.basic.ast

mdedetrich committed:
- Changed representation of JObject for org.json4s.basic.ast from a (String,JValue) to a JField(string:String,value:JValue)

mdedetrich committed:
- Added an extra test for JBoolean

mdedetrich committed:
- Renamed value to isTrue in JBoolean for clarity

mdedetrich committed:
- Changed return type from Option[Boolean], to Some[Boolean], this should theoretically improve performance

mdedetrich committed:
- JBoolean now uses name based extractors, for better performance in Scala 2.11

rossabaker commented:
It's not final yet, but this is a huge step in the right direction, and the growing group can discuss this more effectively from master, and with their own pull requests, than by continuing referencing this ticket.  Merging.
